docker/Dockerfile.base

```diff
@@ -0,0 +1,251 @@
+    # Download the source tarball and verify its integrity.
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#L56
+    && wget -O python.tar.xz "https://www.python.org/ftp/python/$PYTHON3_VERSION/Python-$PYTHON3_VERSION.tar.xz" \
+    && wget -O python.tar.xz.asc "https://www.python.org/ftp/python/$PYTHON3_VERSION/Python-$PYTHON3_VERSION.tar.xz.asc" \
+	&& GNUPGHOME="$(mktemp -d)" \
+    && export GNUPGHOME \
+	&& gpg --recv-keys "$PYTHON_SOURCE_GPG" \
+	&& gpg --verify python.tar.xz.asc python.tar.xz \
+	&& command -v gpgconf > /dev/null && gpgconf --kill all || : \
+	&& rm -rf "$GNUPGHOME" python.tar.xz.asc \
+    # Compile and install Python from source.
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#LL56C2-L99C3
+    && mkdir -p /usr/src/python \
+	&& tar --extract --directory /usr/src/python --strip-components=1 --file python.tar.xz \
+	&& rm python.tar.xz \
+    && cd /usr/src/python \
+    && gnuArch="$(dpkg-architecture --query DEB_BUILD_GNU_TYPE)" \
+    && ./configure \
+        --build="$gnuArch" \
+        --enable-loadable-sqlite-extensions \
+		--enable-optimizations \
+		--enable-option-checking=fatal \
+		--enable-shared \
+		--with-lto \
+		--with-system-expat \
+		--without-ensurepip \
+    && nproc="$(nproc)" \
+	&& LDFLAGS="-Wl,--strip-all" \
+	&& make -j "$nproc" \
+		"EXTRA_CFLAGS=${EXTRA_CFLAGS:-}" \
+		"LDFLAGS=${LDFLAGS:-}" \
+		"PROFILE_TASK=${PROFILE_TASK:-}" \
+    && rm python \
+	&& make -j "$nproc" \
+		"EXTRA_CFLAGS=${EXTRA_CFLAGS:-}" \
+		"LDFLAGS=${LDFLAGS:--Wl},-rpath='\$\$ORIGIN/../lib'" \
+		"PROFILE_TASK=${PROFILE_TASK:-}" \
+		python \
+	&& make install \
+	&& cd / \
+	&& rm -rf /usr/src/python \
+    # Clean up test and cache files.
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#LL100C2-L108C3
+	&& find /usr/local -depth \
+		\( \
+			\( -type d -a \( -name test -o -name tests -o -name idle_test \) \) \
+			-o \( -type f -a \( -name '*.pyc' -o -name '*.pyo' -o -name 'libpython*.a' \) \) \
+		\) -exec rm -rf '{}' + \
+	&& ldconfig \
+    && python3 --version \
+    # Create useful symlinks for python related tools.
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#L124
+	&& for src in idle3 pydoc3 python3 python3-config; do \
+		dst="$(echo "$src" | tr -d 3)"; \
+		[ -s "/usr/local/bin/$src" ]; \
+		[ ! -e "/usr/local/bin/$dst" ]; \
+		ln -svT "$src" "/usr/local/bin/$dst"; \
+	done \
+# ---------------------------------------------------------------------------------------------------------------------
+	# INSTALL PIP AND SETUPTOOLS
+    # https://github.com/docker-library/python/blob/f568f56f28fab0fe87b34db777e2c2861cef002b/3.11/slim-buster/Dockerfile#LL147C2-L166C15
+	&& wget -O get-pip.py "$PYTHON_GET_PIP_URL" \
```

Comment:
It would be better to use the default `pip` module installed with `Python` if possible. You mentioned installing it separately can reduce the final size of the image. Is that significant enough to be worth it?

Reply:
I have just checked the differences. It only comes in around 4MB less when we use the default pip that comes with our Python installation.

Reply:
> I have just checked the differences. It only comes in around 4MB less when we use the default pip that comes with our Python installation.

So is it worth the separate installation?

Reply:
I think the only benefit of doing what we have here is that we ensure that pip hash is checked properly before we install it. Otherwise, 4MB is not worth in terms of image size to walk this path. What do you think about it?